packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,381 @@
+            if (_notProxied(addr)) {
+                continue;
+            }
+
+            vm.etch(addr, code);
+            vm.store(addr, PROXY_ADMIN_ADDRESS, bytes32(uint256(uint160(Predeploys.PROXY_ADMIN))));
+
+            if (_isDefinedPredeploy(addr)) {
+                address implementation = _predeployToCodeNamespace(addr);
+                vm.store(addr, PROXY_IMPLEMENTATION_ADDRESS, bytes32(uint256(uint160(implementation))));
+            }
+        }
+    }
+
+    /// @notice LEGACY_ERC20_ETH is not being predeployed
+    function _setImplementations() internal {
+        _setLegacyMessagePasser();
+        _setDeployerWhitelist();
+        _setWETH9();
+        _setL2StandardBridge();
+        _setL2CrossDomainMessenger();
+        _setSequencerFeeVault();
+        _setOptimismMintableERC20Factory();
+        _setL1BlockNumber();
+        _setGasPriceOracle();
+        _setGovernanceToken();
+        _setL1Block();
+    }
+
+    function _setLegacyMessagePasser() internal {
+        address impl = _predeployToCodeNamespace(Predeploys.LEGACY_MESSAGE_PASSER);
+        vm.etch(
+            impl,
+            vm.getDeployedCode("LegacyMessagePasser.sol:LegacyMessagePasser")
+        );
+
+        _verifyProxyImplementationAddress(Predeploys.LEGACY_MESSAGE_PASSER, impl);
+    }
+
+    function _setDeployerWhitelist() internal {
+        address impl = _predeployToCodeNamespace(Predeploys.DEPLOYER_WHITELIST);
+        vm.etch(
+            _predeployToCodeNamespace(impl),
+            vm.getDeployedCode("DeployerWhitelist.sol:DeployerWhitelist")
+        );
+
+        _verifyProxyImplementationAddress(Predeploys.DEPLOYER_WHITELIST, impl);
+    }
+
+    /// @notice Contract metadata hash appended to deployed bytecode will differ
+    ///         from previous L2 genesis output. Additionally, this contract
+    ///         is NOT proxied.
+    function _setWETH9() internal {
+        vm.etch(
+            Predeploys.WETH9,
+            vm.getDeployedCode("WETH9.sol:WETH9")
+        );
+        vm.store(
+            Predeploys.WETH9,
+            /// string public name
+            hex"0000000000000000000000000000000000000000000000000000000000000000",
+            /// "Wrapped Ether"
+            hex"577261707065642045746865720000000000000000000000000000000000001a"
+        );
+        vm.store(
+            Predeploys.WETH9,
+            /// string public symbol
+            hex"0000000000000000000000000000000000000000000000000000000000000001",
+            /// "WETH"
+            hex"5745544800000000000000000000000000000000000000000000000000000008"
+        );
+        vm.store(
+            Predeploys.WETH9,
+            // uint8 public decimals
+            hex"0000000000000000000000000000000000000000000000000000000000000002",
+            /// 18
+            hex"0000000000000000000000000000000000000000000000000000000000000012"
+        );
+    }
+
+    function _setL2StandardBridge() internal {
```

Comment:
Can you make sure to add natspec for each function? It would be good to comment around the safety of using `getDeployedCode` ie "no side effects in constructor" and "no immutables". If you don't want to copy/paste the same safety invariant, you could number the cases at the top level contract natspec like:

```
The following safety invariants are used when setting state:
1. `vm.getDeployedBytecode` can only be used with `vm.etch` when there are no side effects in the constructor and no immutables in the bytecode
2. A contract must be deployed using the `new` syntax if there are immutables in the code. Any other side effects from the init code besides setting the immutables must be cleaned up afterwards
```

Then you can have a comment on each set function mentioning which invariant it follows

Reply:
Added!

Reply:
Added! Didn't add comments to the post check functions since the called functions are documents - felt redundant